### PR TITLE
18CO Corporation can actually take President's Share

### PR DESCRIPTION
**Pull request by ryandriskel**
Fixes https://github.com/tobymao/18xx/issues/2812 
Fixes https://github.com/tobymao/18xx/issues/2780
Fixes https://github.com/tobymao/18xx/issues/2817 

### Implementation Notes

Looking for a better name for the `corporate_player_share_holders` function if ya got one. It is the set of the corporate share holders and player share holders. For all other current games, it results in the same set as `player_share_holders`

Could also use a better name for `totaling_num`. This function creates an array of shares that have a total percentage matching the number of shares needed to do the swap. This is needed to allow for 20% shares being owned. The default assumes all shares have the same percentage, and thus doesn't even need to consider the share's percentage. The 18CO override finds a bundle totaling the number of shares needed, but swapping 10% shares if the entity owns them.

These changes allow the Corporation to actually obtain the President's certificate, rendering the initial implementation of `find_takeover_corporation` unnecessary. Instead, we can simply check if the corporation is owned by another corporation.

### CODE REVIEW

File: lib/engine/g_18_co/share_pool.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+          shares.combination(n).to_a.select { |b| b.sum(&:percent) == percent }
+        end
+
+        # we want the bundle with the most shares, as higher percent in fewer shares in more valuable
+        matching_bundles.max_by(&:size)
+      end
+
+      def distance(entity_a, entity_b)
+        return 0 if !entity_a || !entity_b
+        return @game.players.size + 485 - entity_b.share_price.price if entity_b.corporation?
```
Comment (tobymao) [inline lib/engine/g_18_co/share_pool.rb:24]: what's 485

Comment (ryandriskel) [reply lib/engine/g_18_co/share_pool.rb:24]: Max share price in 18CO

Comment (ryandriskel) [reply lib/engine/g_18_co/share_pool.rb:24]: Lower overall number is "closer", so this inverts the share price, making higher share value corps have preference.

File: lib/engine/corporation.rb [hunk 2]
```diff
@@ -140,6 +140,10 @@ def share_holders
       @share_holders ||= Hash.new(0)
     end
 
+    def corporate_player_share_holders
```
Comment (tobymao) [inline lib/engine/corporation.rb:143]: remove this method and pass a flag to line 147 player_share_holders(corporations: true) or whatever

File: lib/engine/share_pool.rb [hunk 3]
```diff
@@ -207,10 +207,16 @@ def change_president(presidents_share, swap_to, president)
...
       num_shares = presidents_share.percent / corporation.share_percent
 
-      possible_reorder(president.shares_of(corporation)).take(num_shares).each { |s| move_share(s, swap_to) }
+      totaling_num(possible_reorder(president.shares_of(corporation)), num_shares).each do |s|
+        move_share(s, swap_to)
+      end
       move_share(presidents_share, president)
     end
 
+    def totaling_num(shares, num_shares)
```
Comment (tobymao) [inline lib/engine/share_pool.rb:216]: totaling_num?

Comment (ryandriskel) [reply lib/engine/share_pool.rb:216]: It returns the set of shares that total the number of shares needed for the presidency swap (which is not equal if there are non-10% shares). It is not a boolean.

Comment (tobymao) [reply lib/engine/share_pool.rb:216]: can you rename it ? i still don't really get it

File: lib/engine/g_18_co/share_pool.rb (hunk 1 above)
Comment (ryandriskel) [reply lib/engine/g_18_co/share_pool.rb]: I made it an 18CO constant so its more clear.

File: lib/engine/corporation.rb (hunk 2 above)
Comment (ryandriskel) [reply lib/engine/corporation.rb]: Done

File: lib/engine/share_pool.rb (hunk 3 above)
Comment (ryandriskel) [reply lib/engine/share_pool.rb]: Had to sleep on it to come up with the relevant name.

File: lib/engine/g_18_co/share_pool.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+require_relative '../share_pool'
+
+module Engine
+  module G18CO
+    class SharePool < SharePool
+      def presidency_check_shares(corporation)
+        corporation.player_share_holders(corporate: true)
```
Comment (tobymao) [inline lib/engine/g_18_co/share_pool.rb:9]: this is already default true, why do you use it?

File: lib/engine/share_bundle.rb [hunk 5]
```diff
@@ -50,7 +50,8 @@ def price
     def can_dump?(entity)
       return true unless presidents_share
 
-      (corporation.player_share_holders.reject { |k, _| k == entity }.values.max || 0) >= presidents_share.percent
+      sh = corporation.player_share_holders(corporate: true)
```
Comment (tobymao) [inline lib/engine/share_bundle.rb:53]: this is already default true

Comment (ryandriskel) [reply lib/engine/share_bundle.rb:53]: Good catch. Its should not be the default, that's why.

**Review by tobymao — APPROVED** [review]

**Pull request merged**